Define default Python engine by system and user

mmisol opened the pull request:
### Purpose

Adds the ability to define a default Python engine for the system. This
is done through a new property in the model start configuration, named
'DefaultPythonEngine'.

Also adds the ability for users to define through a setting their own
default. This can be set to either engine or to fallback to the system
default, whatever it is.

Serialization is tweaked so that 'IronPython2' remains the
deserialization default but is not the absolute default. Engine will be
unspecified on node creation but become loaded as soon as the property
is accessed, to avoid errors on automatic execution.

In order to make system and user settings available on the node, and
avoiding further changes to the API, the properties were made static
and internal.

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [x] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [x] This PR modifies some build requirements and the readme is updated

### Reviewers

@mjkkirschner @QilongTang 

### FYIs

@DynamoDS/dynamo

mmisol commented:
Just wanted to make it clear this is a Work In Progress. Please ignore the commented out code. The idea is to get some feedback on how this looks before wrapping up, since there are several questionable decisions made.

mmisol commented on src/DynamoApplications/StartupUtils.cs:
```diff
@@ -109,8 +111,9 @@ public static CommandLineArguments Parse(string[] args)
-                Add("gp=|GP=|geometrypath=|GeometryPath=", "relative or absolute path to a directory containing ASM. When supplied, instead of searching the hard disk for ASM, it will be loaded directly from this path.", gp => asmPath = gp);
+                " - if you wish to import multiple dlls - use this flag multiple times: -i 'assembly1.dll' -i 'assembly2.dll' ", i => importPaths.Add(i))
+                .Add("gp=|GP=|geometrypath=|GeometryPath=", "relative or absolute path to a directory containing ASM. When supplied, instead of searching the hard disk for ASM, it will be loaded directly from this path.", gp => asmPath = gp);
+                //.Add("py=|PY=|pythonEngine=|PythonEngine=", "Default Python script engine", py => pythonEngine = py);
```
At some point I thought this could also be a command line option, but then disregarded that. Integrators won't use that and not sure we would ever use it for Sandbox. It's also not fully implemented.

mmisol commented on src/DynamoCore/Graph/Nodes/NodeModel.cs:
```diff
@@ -2515,6 +2515,11 @@ public static NodeMigrationData MigrateShortestLacingToAutoLacing(NodeMigrationD
+        public virtual void OnBeforeAdd(Dynamo.Models.DynamoModel dynamoModel, WorkspaceModel workspace)
+        {
+            // Purposely empty. For inheritors to override if needed.
+        }
```
In my mind it made sense to add this sort of life-cycle method to the node. It allowed to act at the right time and provides the required `DynamoModel`.

mmisol commented on src/DynamoCore/Models/DynamoModel.cs:
```diff
@@ -501,6 +511,14 @@ public interface IStartConfiguration
+            /// <summary>
+            /// Provide values for startup options
+            /// </summary>
+            //IDictionary<string, string> StartupOptions { get; set; }
```
This is like an alternative API I had in mind at first. It would allow to provide command-line options in a generic way. The only benefit I can think about it is that we don't add an explicit Python property to the config. On the other hand, it is messier to add entries to a dictionary than assigning properties.

mjkkirschner replied:
I've always thought it would be nice to be able to override the entire config with an xml or json file - but thats a discussion for another time I think.

Dewb replied:
We already have multiple, partially overlapping systems to manage config, with StartConfiguration, IPreferences, StartupUtils.CommandLineArguments, properties copied into models, etc. Rationalizing all of this should be a priority for the next API-breaking release, would love someone to research and draft a proposal.

mmisol commented on src/DynamoCoreWpf/UI/Converters.cs:
```diff
@@ -1108,6 +1108,48 @@ public object ConvertBack(object value, Type targetType, object parameter, Syste
+    //        throw new NotImplementedException();
+    //    }
+    //}
+
```
Some experiments I did with WPF before ending up with the current UI.

mmisol commented on src/DynamoCoreWpf/Views/Core/DynamoView.xaml:
```diff
@@ -724,6 +724,55 @@
+                        <MenuItem Focusable="False"
+                                  Header="{x:Static p:Resources.DynamoViewSettingsDefaultPythonEngine}">
+                            <ComboBox ItemsSource="{Binding DefaultPythonEngineOptions}" SelectedValue="{Binding DefaultPythonEngine, Mode=TwoWay}" SelectedValuePath="Value" DisplayMemberPath="Name"/>
+                        </MenuItem>
```
Used a drop down. As you can see from the commented markup I also tried child checkable menu items, which had some limitations. Open to discussion.

QilongTang replied:
@mmisol Looking back on this, what limitations are in this case? Is it because we want to be dynamic on this and not achievable using child menu items?

mmisol commented on src/Libraries/PythonNodeModels/PythonNode.cs:
```diff
@@ -29,13 +32,15 @@ public PythonCodeMigrationEventArgs(string oldCode, string newCode)
     public abstract class PythonNodeBase : VariableInputNode
     {
-        private PythonEngineVersion engine = PythonEngineVersion.IronPython2;
+        private PythonEngineVersion? engine = null;
```
The engine became nullable, as it cannot be initialized at this point. We have no access to `DynamoModel`. Luckily JSON serialization handles this change without problems.

mmisol commented on src/Libraries/PythonNodeModels/PythonNode.cs:
```diff
@@ -54,6 +59,30 @@ protected PythonNodeBase()
+                {
+                    // In the absence of both a setting and system default, default to deserialization default.
+                    Engine = PythonEngineVersion.IronPython2;
+                }
```
user setting > system default > deserialization default

mjkkirschner commented on src/DynamoCore/Configuration/PreferenceSettings.cs:
```diff
@@ -333,6 +333,11 @@ public string PythonTemplateFilePath
         public bool ShowTabsAndSpacesInScriptEditor { get; set; }
 
+        /// <summary>
+        /// Engine used by default for new Python script and string nodes. If defined this takes precedence over any system settings.
```
do you mean if not null? Does empty string qualify as defined?

mmisol replied:
I'm pretty sure I'm using empty string for not defined.

mjkkirschner replied:
can you document that in this comment?

mjkkirschner commented on src/DynamoCore/Graph/Nodes/NodeModel.cs:
```diff
@@ -2515,6 +2515,11 @@ public static NodeMigrationData MigrateShortestLacingToAutoLacing(NodeMigrationD
 
         #endregion
 
+        public virtual void OnBeforeAdd(Dynamo.Models.DynamoModel dynamoModel, WorkspaceModel workspace)
```
this is quite a big change - in the rest of Dynamo it's common for models to only have references to their children - ie DynamoModel knows about NodeModels - and NodeModels know about portModels, but not the other way around.

Curious why you need this reference - maybe it will become clear.

mjkkirschner replied:
hmm - it makes a lot of sense that the python nodeModel needs to continue working even in headless Dynamo.

mjkkirschner replied:
I'm curious what you think of using the active `ExecutionSession` instead - storing the pythonEngine there and extracting it from the static map when the node is constructed.

mmisol replied:
I'm open to change this. `ExecutingSession` is looking uglier each time we add another new thing to it, but I guess it can be used.

This does expose more data, but we are not being super consistent with the visibility anyway. For view models the rules are pretty different than what you mention here.

I don't think this would break headless Dynamo, unless you added a node to a graph dynamically (can you do that headless?) and forgot to assign an `Engine` to it.

mjkkirschner replied:
you can add a node to a graph dynamically in headless mode for sure - consider some headless client of Dynamo that calls the createNodeCommand - but - I was actually trying to make the point that I thought your solution was better than my original suggestion because this doesn't require the NodeView to be around to correctly set the engine.

mjkkirschner replied:
for some context - I believe historically, there was a giant static class called something like `DynamoSettings` or `DynamoModelEntryPoint` that was used from everywhere, this made the separation between core and wpf layers very difficult to enforce, and a big refactor was undertaken.

There was also the idea that one day we'd have multiple DynamoModels running at the same time in the same process, but this hit other roadblocks.

I'm not sure the pain for nodeModel developers (like you here) is worth the old ideas from that refactor any longer.

mmisol replied:
Multiple DynamoModel sounds intriguing. In any case a node would belong to just one of those, right? 

Giving a node model access to its parent workspace model and that one having access to its parent dynamo model does not sound so bad to me. It feels having access to the context a node model is in should be more direct than relying on global classes. 

In any case, we probably don't want to make a refactor part of this PR. I'll stick to `ExecutingSession` for now.

mmisol replied:
Now thinking again from a security perspective, accessing the `DynamoModel` opens the door for a node to do a bunch of things we may not want them to do. Note that the risk currently exists for a custom `NodeView` as they can access the `DynamoViewModel` and then to the `DynamoModel`.

It might be a  good idea to think this further and consider refactoring in 3.0.

mjkkirschner commented on src/DynamoCore/Models/DynamoModel.cs:
```diff
@@ -501,6 +511,14 @@ public interface IStartConfiguration
+            /// <summary>
+            /// Default Python script engine
+            /// </summary>
+            string DefaultPythonEngine { get; set; }
```
modifying this interface is an API break - we can add a new interface instead, and merge them together in 3.x

mmisol replied:
An alternative can be not have this in the interface but keep it in `DefaultStartConfiguration`, and cast and see in `DynamoModel` constructor. I see some other properties on this boat already.

mjkkirschner commented on src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:
```diff
@@ -46,6 +47,20 @@ public interface IDynamoViewModel : INotifyPropertyChanged
         ObservableCollection<WorkspaceViewModel> Workspaces { get; set; } 
     }
 
+    public class SimpleViewModel<T> : ViewModelBase, INotifyPropertyChanged
```
experiment?

mjkkirschner replied:
apparently you can use old style tuple items as bindings - you cannot use new style value tuples though as those members are fields - anyhow, something I consider I guess.

mjkkirschner replied:
https://stackoverflow.com/questions/43208852/is-it-possible-to-bind-to-a-valuetuple-field-in-wpf-with-c7

mmisol replied:
Maybe not the right place for it, but I thought this could be a good way to to it. I'm not a fan of tuples, specially because the property names are positional and have no meaning.

mjkkirschner commented on src/Libraries/PythonNodeModels/PythonNode.cs:
```diff
@@ -29,13 +32,15 @@ public PythonCodeMigrationEventArgs(string oldCode, string newCode)
         /// Return the user selected python engine enum.
         /// </summary>
-        public PythonEngineVersion Engine
+        public PythonEngineVersion? Engine
```
I wonder if this qualifies as an API break...

mmisol replied:
It does.

mjkkirschner replied:
that seems like a problem ;) Can we add a new enum like none?

mmisol replied:
I can do this. I just felt we weren't tied to this API yet since it was recently added, but I guess people may be using it already.

mjkkirschner commented on src/Libraries/PythonNodeModels/PythonNode.cs:
```diff
@@ -88,7 +117,7 @@ protected override string GetInputTooltip(int index)
             // Here we switched to use the AstFactory.BuildFunctionCall version that accept
             // class name and function name. They will be set by PythonEngineSelector by the engine value. 
-            PythonEngineSelector.Instance.GetEvaluatorInfo(Engine, out string evaluatorClass, out string evaluationMethod);
+            PythonEngineSelector.Instance.GetEvaluatorInfo(Engine.Value, out string evaluatorClass, out string evaluationMethod);
```
I guess so?

mjkkirschner commented on src/Libraries/PythonNodeModels/PythonNode.cs:
```diff
@@ -54,6 +59,30 @@ protected PythonNodeBase()
             ArgumentLacing = LacingStrategy.Disabled;
         }
 
+        public override void OnBeforeAdd(DynamoModel dynamoModel, WorkspaceModel workspace)
```
workspace looks unused - just thinking ahead?

mmisol replied:
In my mind, this "life-cycle method" made more sense if you only got the workspace where the node was being added. However, you have the same problem, `WorkspaceModel` has no visibility of the `DynamoModel` associated with it. In the end I kept both because only providing the `DynamoModel` made less sense as an API, but it's true it's not really used.

QilongTang commented on src/DynamoCore/Configuration/PreferenceSettings.cs:
```diff
@@ -388,6 +393,7 @@ public PreferenceSettings()
             PythonTemplateFilePath = "";
             IsIronPythonDialogDisabled = false;
             ShowTabsAndSpacesInScriptEditor = false;
+            DefaultPythonEngine = string.Empty;
```
Do we want to make PythonEngineVersion enum static so we can refer anywhere with known value options?

mmisol replied:
Do you mean make the `DefaultPythonEngine` preference setting static? I guess that could work, but we also have the system setting, so we'd have to make them both static I guess. It could be an alternative to putting stuff into `ExecutingSession`, which may be cleaner if we can use `InternalsVisibleTo`. Will check that out.

mmisol commented:
Based on the feedback, I went with internal static properties and got rid of `OnBeforeAdd`. Also cleaned up and added tests. Please take another look @mjkkirschner @QilongTang

mmisol commented:
Tests are passing now https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/pullRequestValidation/1284/

mjkkirschner commented on src/DynamoCore/Models/DynamoModel.cs:
```diff
@@ -423,6 +425,14 @@ public IEnumerable<WorkspaceModel> Workspaces
+        {
+            get
+            {
+                return defaultPythonEngine;
```
just curious, why is this not using an auto property?

mmisol replied:
Can use with private set. Will change

mjkkirschner commented on src/DynamoCore/Models/DynamoModel.cs:
```diff
@@ -553,6 +567,13 @@ public static DynamoModel Start(IStartConfiguration configuration)
         {
+            if (config is DefaultStartConfiguration)
+            {
+                // This is not exposed in IStartConfiguration to avoid a breaking change.
```
do you want to add a TODO?

mmisol replied:
The next line is a `TODO` without a prefix, so I'll add it there.

QilongTang replied:
Thanks for adding it @mmisol , can you mark it specially for 3.0? A bit easier to track all the 3.0 issues in code base

mmisol replied:
Will do 👍

mjkkirschner commented on src/DynamoCore/Models/DynamoModel.cs:
```diff
@@ -553,6 +567,13 @@ public static DynamoModel Start(IStartConfiguration configuration)
+            {
+                // This is not exposed in IStartConfiguration to avoid a breaking change.
+                // This fact should probably be revisited in next major.
+                defaultPythonEngine = ((DefaultStartConfiguration)config).DefaultPythonEngine;
```
would it ever happen that we've loaded an old xml config and this will throw?

mmisol replied:
I don't know in which context this gets serialized/deserialized. XML serialization OOTB does not throw in cases like this but I don't know what you are doing exactly. Can you point to where that happens?

mjkkirschner commented on src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:
```diff
@@ -67,7 +68,13 @@ public int ScaleFactorLog
-        
+
+        private ObservableCollection<SimpleViewModel<string>> defaultPythonEngineOptions = new ObservableCollection<SimpleViewModel<string>>();
+        public ObservableCollection<SimpleViewModel<string>> DefaultPythonEngineOptions
```
why the need for the field?

mmisol replied:
I didn't want to initialize this more than once.

mjkkirschner commented on src/DynamoCoreWpf/ViewModels/Core/SimpleViewModel.cs:
```diff
@@ -0,0 +1,31 @@
+    /// Dead simple view model for when you only need to have a different display name and underlying value.
+    /// </summary>
+    /// <typeparam name="T">The type of your value</typeparam>
+    public class SimpleViewModel<T> : ViewModelBase, INotifyPropertyChanged
```
do you expect this to ever grow to contain more properties?

Maybe we can call it `NameValuePairViewModel`?

mmisol replied:
Ok

mjkkirschner commented on src/DynamoCoreWpf/ViewModels/ViewModelBase.cs:
```diff
@@ -20,15 +20,6 @@ public bool IsDebugBuild
         }
 
-        [JsonIgnore]
-        public bool IsPython2ObsoleteDebugModeEnabled
```
why is this removed?

mmisol replied:
That's a good question. I saw it was not being used. You could say removing it is a breaking change, but it was added a couple of weeks ago, probably by mistake.

mjkkirschner commented on src/Libraries/PythonNodeModels/PythonNode.cs:
```diff
@@ -29,15 +32,25 @@ public PythonCodeMigrationEventArgs(string oldCode, string newCode)
 
         [JsonConverter(typeof(StringEnumConverter))]
+        [JsonProperty(DefaultValueHandling = DefaultValueHandling.Populate)]
+        [DefaultValue("IronPython2")]
```
can we use nameof?

mmisol replied:
Yes

mjkkirschner commented on src/Libraries/PythonNodeModels/PythonEngineSelector .cs:
```diff
@@ -9,6 +9,7 @@ namespace PythonNodeModels
     /// </summary>
     public enum PythonEngineVersion
     {
+        Unspecified,
```
this should go at the end of the list as this change re-orders all the existing enums - ie:
Ironpython2 used to = `0`, but after this change it is `1`.

mmisol replied:
Technically this only breaks binary compatibility. I put this in the balance with following best practices, and opted for the latter. Let me know if you'd rather change this anyway.

mjkkirschner replied:
my experience with this type of change is pretty complex confusion - code which was previously compiled against Dynamo 2.7 and selected engine ironPython2 will now be selecting unspecified.

I guess it's unlikely that any clients besides our own would ever be in this situation though.

If you've already thought about it then I'm fine with your decision. 👍

mjkkirschner commented on src/Libraries/PythonNodeModels/PythonNode.cs:
```diff
@@ -29,15 +32,25 @@ public PythonCodeMigrationEventArgs(string oldCode, string newCode)
+            get
+            {
+                // This is a first-time case for newly created nodes only
+                if (engine == PythonEngineVersion.Unspecified)
```
👍 , but please see comment about enum order

mjkkirschner commented on src/Libraries/PythonNodeModels/PythonNode.cs:
```diff
@@ -48,6 +61,47 @@ public PythonEngineVersion Engine
+        /// <summary>
+        /// Available Python engines.
+        /// </summary>
+        public ObservableCollection<PythonEngineVersion> AvailableEngines
```
same question about auto properties

mmisol replied:
This is for the purpose of initializing it only once.

mjkkirschner replied:
I guess I'm still a bit confused by this answer - 
can't you do `Public ObserverableCollection<PythonEngineVersion> {get; set;} = ObserverableCollection<PythonEngineVersion>(){ yada yada}` ?

mmisol replied:
Yes, you are right. I guess I don't really like to initialize in that way when it takes more than one line. If it was just creating the collection I wouldn't mind.

mmisol replied:
Oh, I see you put `{ yada yada}` in same line. I don't really like that either 😄

mjkkirschner commented on test/DynamoCoreWpfTests/PythonNodeCustomizationTests.cs:
```diff
@@ -15,7 +15,7 @@ namespace DynamoCoreWpfTests
     public class PythonNodeCustomizationTests : DynamoTestUIBase
     {
-        private readonly List<string> expectedEngineMenuItems = Enum.GetNames(typeof(PythonNodeModels.PythonEngineVersion)).ToList();
+        private readonly List<string> expectedEngineMenuItems = Enum.GetNames(typeof(PythonNodeModels.PythonEngineVersion)).Skip(1).ToList();
```
? comment on this maybe?

mmisol replied:
I think I'd rather change the code to make it more clear.

QilongTang commented on src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:
```diff
@@ -583,6 +604,11 @@ public static DynamoViewModel Start(StartConfiguration startConfiguration = new
+            // These options are currently fixed. Eventually we should make them dynamic I guess.
+            defaultPythonEngineOptions.Add(new NameValuePairViewModel<string>(string.Empty, WpfResources.DefaultPythonEngineNone));
+            defaultPythonEngineOptions.Add(new NameValuePairViewModel<string>("IronPython2", "IronPython2"));
+            defaultPythonEngineOptions.Add(new NameValuePairViewModel<string>("CPython3", "CPython3"));
```
Sry to get to it late, this is the reason why I mentioned earlier if we want to just use the enums we already exposing as `PythonEngineVersion`, we just need to add a default entry

QilongTang replied:
I guess what I mean is instead of defining a fix collection but loop through the enums and construct it to reduce defined strings. Not sure if you have thought about it already

mmisol replied:
The thing is that while `IronPython2` and `CPython3` are engine names and can go as they are, `Unspecified` is the absence of an engine and should be translated. Probably in different ways according to the context.
Also, not sure how permanent our enum is going to live forever. If we are going to open up the possibility for new engines, we'll need a solution based on names/strings.

mmisol commented:
@QilongTang If you are ok I'll go ahead and merge this.

QilongTang commented:
@mmisol I am OK to merge. Since this piece of work will eventually go into D4R and Civil3D. I would like us do some testing and keep it in mind that we may revisit some details here. I like the fact we are using preference setting, just seems more manageable to me. Good work!
